Explicit CoordinateSequence support for XYZM

jodygarnett opened the pull request:
Checklist:

* [x] CoordinateSequence.getMeasures() default method
* [x] Update to PackedCoordinateSequence
* [x] discussion and feedback
* [x] Introduce Coordinate accessors for getX(), getY(), getZ(), getM()
* [x] Introduce explicit CoordinateXY, CoordinateXYM, CoordinateXYZM
* [x] Updated CoordinateTest to cover subclasses
* [x] Added CoordinateSequence hasZ and hasM default methods to perform common dimension and measures checks
* [x] Updated PackedCoordinateSequenceTest to try out XY, XYM, XYZ, XYZM functionality
* [x] Updated CoordinateArraySequence as well
* [x] Added CoordinateSequence.createCoordinate() default method, to assist with client use
* [x] Reworked ExtendedCoordinateSequence example

Ideally:

* [ ] proof of fitness with feedback from https://github.com/locationtech/jts/pull/291 WKT Reader/Writer enhancement

jodygarnett commented:
So reading some of the javadocs indicates that it is up to the client code to assign a meaning to dimensions beyond 2.

Son in GeoTools if we have a SRS with dimension 2 and a geometry with dimension 3 we would know we are looking at XYM data.

jodygarnett commented on modules/core/src/main/java/org/locationtech/jts/geom/CoordinateSequence.java:
```diff
@@ -60,6 +60,14 @@
+   * Returns the number of measures in each coordinate for this sequence.
+   * @return
+   */
+  default int getNumberOfMeasures() {
```
here is the default interface

bjornharrtell commented:
Not sure, should be possible to fix in case of troubles, so don't let it be a blocker.

basisbit commented:
could you please specify what Z and M should represent? I guess for Z this is obvious for many use cases, some clarification would still be nice.

FObermaier commented:
Does the introduction of `getNumberOfMeasures` imply changing the meaning of `getDimensions` (as in GeoTools' CoordinateAccess interface? If so this would impose a breaking change for  `CoordinateSequence` implementations like `PackedCoordinateSequence` that are used with a dimension greater than 3.

dr-jts commented:
@FObermaier my thoughts on answers to your questions:

* the semantics of `CoordinateSequence.getDimension` should not change. It should continue to return the total number of ordinates available.
* I think @jodygarnett  meant `CoordinateSequence.getZ`.  This returns NaN if no Z dimension was defined.  
* Again, I think the suggestion is to add `CoordinateSequence.getM`.  

The Z and M accessors are suggested in order to make the support for Z and M explicit in the API (and to match the explicit support in the SFSQL spec).  

It might be good to provide a `CoordinateSequence.hasZ` method, in order to encapsulate the logic of determining this (and this can be made a default interface method).  `hasM `could be provided as well, but does have the fairly trivial equivalent of `getNumberOfMeasures == 1`

jodygarnett commented:
Guess I am always confused if M counts as a ordinate with respect to getDimension().

dr-jts commented:
Yeah, it's confusing.  In JTS dimension has always referred to the total number of ordinates available.  If we change that not sure what will happen...  

If `getDimension `is changed to be just the number of geometric dimensions, then pretty sure that a `getNumberOfOrdinates `should be available as well, for ease of iteration.

dr-jts commented:
Further to the question of the semantics of getDimension, my read of the SFSQL (below) is that getDimension should be the total number of ordinates, including the M value(s) if present, based on the second sentence.

> COORD_DIMENSION — the number of ordinates used in the complex, usually corresponds to the number of dimensions in the spatial reference system. **If an "M" ordinate is included it shall be one greater than the number of dimensions of the spatial reference system**.

jodygarnett commented:
You can see my quote above about GEOMETRY_TYPE having different values for POINT POINTZ POINTM POINTZM. If we went that approach and reported back the correct SFSQL type .. it would be a pain and not be visible at the CoordianteSequence level.

I would like to keep getDimension() in agreement with the SFSQL spec if we can:

> dimension = spatialDimension + numberOfOrdinates 

Would rather see an approach using numbers, rather than hasM() and hasZ() methods .. to allow for XYZMN etc...

dr-jts commented:
I don't understand `dimension = spatialDimension + numberOfOrdinates`. What is `numberOfOrdinates `?

Agreed that a numeric approach is needed (as you say, to allow for measures > 1).  But doesn't preclude  having the predicates too.

jodygarnett commented:
@dr-jts and @jnh5y can I ask for a review, I have responded I think to most of the feedback.

I ended up having CoordianteSequence.getDimension() as described in the conversation above, and CoordianteSequence.getMeasures() being the number of measures (so if we have XYZM dimension is 4 and measures is 1.

I also removed the mutability from PackedCoordianteSequence as the factory methods were provided, and there was no need for PackagedCoordinateSequence to hold on to state (especially when their were static constances for PackagedCoordinateSequence.DOUBLE and PackagedCoordinateSequence.FLOAD.

@jnh5y I have introduced the specific Coordiante subclasses as you requested, so we end up with:

* Coordinate - stores XYZ, the getM() method always returns Double.NaN
* CoordinateXY - both getZ() and getM() return Double.NaN. The field "z" is still visible from superclass but not used
* CoordinateXYM
* CoordinateXYZM

elahrvivaz commented on modules/core/src/main/java/org/locationtech/jts/geom/Coordinate.java:
```diff
@@ -47,37 +48,55 @@
+  /**
+   * The x-ordinate.
    */
   public double x;
```
would it make sense to deprecate direct access to x and y as well? It seems potentially confusing to have `coord.x`, `coord.y` and `coord.getZ()` be the established access pattern

jodygarnett replied:
I would like to do so, but discussion with @jnh5y and @dr-jts has me considering if having explicit classes, with access to the primitives, is required. Martin quite correctly focuses on these being primary a data structures, rather than an object.

With java value types on the horizon we are likely to change to the `struct` idea from C. Really want to have coordinates represent a block of [xy], [xyz], [xym], [xyzm] if continuous memory, so when we make an array of them everything is co-located in memory.

jodygarnett replied:
So my design hat agrees with you, prefer getX() and getY(), make them final so the JRE can inline them to avoid performance penalty which is a key concern ...

dr-jts replied:
Have to keep public access to Coordinate x and y fields for backwards compatibility.  These could be removed in a future non-compatible version (i,e, JTS 2)

dr-jts replied:
Well, unless a big refactoring is done (as I see @jodygarnett has already done for the z field).

jodygarnett replied:
just because refactor is done for JTS codebase, does not mean I want to accidentally break client code.

jodygarnett replied:
I really am not sure value types is on the java roadmap ...

dr-jts replied:
The problem with marking as Deprecated is that that triggers compiler errors, which can be annoying.  Can comment them as deprecated, though.

jodygarnett replied:
really? I thought it just made compiler warnings?

From the annotation javadocs:

> A program element annotated &#64;Deprecated is one that programmers
> are discouraged from using, typically because it is dangerous,
> or because a better alternative exists.  Compilers warn when a
> deprecated program element is used or overridden in non-deprecated code.

In anycase I have it commended it as deprecated as the javadoc approach allows a message to be provided

dr-jts replied:
Well, yes, warnings not errors.  Still, it's nice to allow compiling warning-free.

elahrvivaz replied:
my $0.02 - probably a good idea to comment and also mark them. the compiler warnings are useful to raise awareness that users should be moving away from the methods, whereas most wouldn't notice a comment change and then face an unexpected compile failure whenever they actually go away.

dsmiley replied:
LOL emitting warnings so people change their code is entirely the point.  Then they change their code and the warnings go away

dr-jts replied:
I have been in situations where compile-time warnings trigger a rejection of the update to the library, and it is not possible to change the client code.  So I'm very wary of deprecating things until at least a few releases containing the alternative functionality.

elahrvivaz commented on modules/core/src/main/java/org/locationtech/jts/geom/CoordinateSequence.java:
```diff
@@ -111,15 +145,48 @@
+   */
+  default double getM(int index) {
+     if( getDimension() > 2 && getMeasures() > 0 ) {
+	 final int mIndex = getDimension()-getDimension();
```
should this be `getDimension() - getMeasures()`?

jodygarnett replied:
yes, yes it should. And I should have test cases before this goes live :)

elahrvivaz commented on modules/core/src/main/java/org/locationtech/jts/geom/Coordinates.java:
```diff
@@ -0,0 +1,93 @@
+	    return 2;
+	}
+	else if(coordiante instanceof CoordinateXYM) {
+	    return 2;
```
should this be 3?

elahrvivaz commented on modules/core/src/main/java/org/locationtech/jts/geom/impl/PackedCoordinateSequence.java:
```diff
@@ -250,19 +270,23 @@ public Double(float[] coordinates, int dimensions) {
-          coords[i * this.dimension + 2] = coordinates[i].z;
+          coords[i * this.dimension + 2] = coordinates[i].getOrdinate(2); // Z or M
+        if (this.dimension >= 4)
+            coords[i * this.dimension + 2] = coordinates[i].getOrdinate(3); // M
```
should this be `+ 3`

jodygarnett commented:
Okay folks test case written, this has collected all feedback provided.

jratike80 commented:
I suppose we do not want to read literally the citation from SFSQL because it seems to say that for example EPSG:4326 that is Ellipsoidal 2D CS it is not allowed to use XYZ or XYZM coordinates while XYM is OK with COORD_DIMENSION=3.

> COORD_DIMENSION — the number of ordinates used in the complex, usually corresponds to the number of dimensions in the spatial reference system. If an "M" ordinate is included it shall be one greater than the number of dimensions of the spatial reference system.

jratike80 commented:
What is supposed to happen when XY(Z)M coordinates are passed to JTS operations which likely destroy the meaning of measures? For example, most often M is used for linear referencing and M does not make sense if geometries are buffered.

dr-jts commented on modules/core/src/main/java/org/locationtech/jts/geom/CoordinateSequence.java:
```diff
@@ -1,5 +1,5 @@
 /*
- * Copyright (c) 2016 Vivid Solutions.
+ * Copyright (c) 2018 Vivid Solutions, and others.
```
I don't think this is a valid copyright?

jodygarnett replied:
It is the format used by eclipse source code so I figured it was good:


![](http://www.codejava.net/images/articles/ides/eclipse/copyright/Copyright_Tool_default_setting.png)

dr-jts commented on modules/core/src/main/java/org/locationtech/jts/geom/CoordinateSequence.java:
```diff
@@ -45,22 +45,76 @@
+   *
+   * @return the number of measures included in dimension
+   */
+  default int getMeasures() {
```
This might be better named getNumberOfMeasures, or getMeasureCount

jodygarnett replied:
I settled on shorter, after trying getNumberOfMeasures() initially.

jnh5y replied:
Yeah, I'll say that I'd expect getMeasures to return an Array/List of the measures rather than a count.  Can we change the name?

dr-jts commented:
What are the use cases for CoordinateXY, CoordinateXYM, etc?  

Internally JTS just uses Coordinate as a intermediate object to allow easy computing and reference to coordinate values.  This can be served by Coordinate (and perhaps Coordinate with an added M value if and when measure-based algorithms are provided).

Otherwise it is encourage to just use the CoordinateSequence API.

If the other classes are not used, then perhaps there is no need to provide them?

Another issue is that they are hard to use programmatically, since the dimension is statically encoded in the class.

jodygarnett commented:
jratike80 we found inconsistent descriptions on the meaning of dimension (both including "m" and excluding "m" as described in your quote). We decided to go with including M, so that dimension describes the limit accepted by the getOrdiante method, if not client code is always stuck doing math...

jodygarnett commented:
@dr-jts the use case for CoordinateXY, CoordinateXYM etc... is from @jnh5y who wanted explicit implementations to better control memory use. If we had CoordinateXY with only an x and a y field it would be ideal, or as an interface to produce flyweight objects for CoordinateSequence access.

jnh5y commented on modules/core/src/main/java/org/locationtech/jts/geom/CoordinateSequence.java:
```diff
@@ -45,22 +45,76 @@
+   * @return true if {@link #getZ(int)} is supported.
+   */
+  default boolean hasZ() {
+      return (getDimension()-getMeasures()) > 2;
```
Is there a situation where getDimension()-getMeasures() >= 2?

jnh5y commented on modules/core/src/main/java/org/locationtech/jts/geom/CoordinateSequence.java:
```diff
@@ -45,22 +45,76 @@
+   * @return true if {@link #getM(int)} is supported.
+   */
+  default boolean hasM() {
+      return getDimension()>2 && getMeasures() > 0;
```
@jodygarnett is there any used case for 1-d geometries with a measure?  Is checking getMeasures() > 0 sufficient?

jnh5y commented on modules/core/src/main/java/org/locationtech/jts/geom/CoordinateSequence.java:
```diff
@@ -111,15 +165,49 @@
    *
    * @param index  the coordinate index in the sequence
-   * @param ordinateIndex the ordinate index in the coordinate (in range [0, dimension-1])
+   * @param ordinateIndex the ordinate index in the coordinate (in range [0, dimension-numberOfMeasures-1])
```
I'm not sure about the range here.  It seems to me that the range would still be [0, dimension-1].  

The bit from [0, dimension-numberOfMeasures-1] would be the XY(Z) values.  For numberOfMeasures > 0, the bit from [dimension-numberOfMeasures, dimension-1] would the measures.

jnh5y replied:
Should you update this line of documentation then?

jnh5y commented on modules/core/src/main/java/org/locationtech/jts/geom/CoordinateSequence.java:
```diff
@@ -133,7 +221,7 @@
    *
    * @param index  the coordinate index in the sequence
-   * @param ordinateIndex the ordinate index in the coordinate (in range [0, dimension-1])
+   * @param ordinateIndex the ordinate index in the coordinate (in range [0, dimension-numberOfMeasures1])
```
numberOfMeasures instead of numberOfMeasures1

jnh5y replied:
Remaining typo.

basisbit commented on modules/core/src/main/java/org/locationtech/jts/geom/Coordinate.java:
```diff
@@ -1,5 +1,5 @@
 /*
- * Copyright (c) 2016 Vivid Solutions.
+ * Copyright (c) 2018 Vivid Solutions, and others.
```
how about `Copyright (c) 2018 Vivid Solutions, and contributors.`? That is what most projects do and you don't have the overhead of contacting every previous contributor and hope they sign a copyleft contract.

jodygarnett replied:
I don't mind either way, "and others" was how the eclipse projects were doing it.

Really we can ask @jnh5y as a locaitontech committer representative to look this up for us and get back with an answer.

jnh5y replied:
From looking at https://www.eclipse.org/projects/handbook/#ip-copyright-headers, it looks like "and Contributors to the Eclipse Foundation" may be acceptable.  (I'd be comfortable with that since it is listed in the Eclipse documentation.)

jodygarnett replied:
I am just reverting it, we can sort out headers another time without delaying this PR.

jnh5y commented on modules/core/src/main/java/org/locationtech/jts/geom/CoordinateXY.java:
```diff
@@ -0,0 +1,98 @@
+      case X: return x;
+      case Y: return y;
+      }
+      throw new IllegalArgumentException("Invalid ordinate index: " + ordinateIndex);
```
Should this be the 'default' case in the switch?

jodygarnett replied:
that would also be a valid way to represent it.

jnh5y commented on modules/core/src/main/java/org/locationtech/jts/geom/CoordinateXYZM.java:
```diff
@@ -0,0 +1,96 @@
+    }
+  }
+  
+  public String toString() {
```
On the topic of the 'toString' methods...  It seems that the original Coordinate.toString has ()'s.  Should we be careful here?

jodygarnett replied:
So original is Coordinate:

```
    return "(" + x + ", " + y + ", " + z + ")";
```

And ExtendedCoordiante:

```
    String stringRep = x + " " + y + " m=" + m;
    return stringRep;
```

I am happy to go with '(', kind of like the use of 'm=' so it is clear when measures are in play.

jnh5y replied:
+1, I like it.

jnh5y commented on modules/core/src/main/java/org/locationtech/jts/geom/Coordinates.java:
```diff
@@ -0,0 +1,90 @@
+    } else if (dimension == 4 && measures == 1) {
+      return new CoordinateXYZM();
+    }
+    return new Coordinate();
```
Hmm...  This seems a tough case to handle properly...

jodygarnett replied:
I think I should make a CoordinateSequence.createCoordiante() method, it will make the ExtendedCoordianteSequence example easier to work with as well. If not we are demanding that these subclasses be used which is poor design.

jnh5y commented on modules/core/src/main/java/org/locationtech/jts/geom/impl/CoordinateArraySequence.java:
```diff
@@ -153,7 +201,12 @@ public Coordinate getCoordinateCopy(int i) {
   public void getCoordinate(int index, Coordinate coord) {
```
As a sanity check, this seems like an easy place for a user to do something silly.  If they pass in the wrong kind of Coordinate, they could get a no-Z/M-support error message, right?  

In JTS 2.0, it'd be great to see if we can be clever enough to handle this at the type-system.

jodygarnett replied:
In this case we trust the `coord` set methods to handle things appropriately.

jnh5y commented on modules/core/src/main/java/org/locationtech/jts/geom/impl/CoordinateArraySequence.java:
```diff
@@ -170,6 +223,31 @@ public double getY(int index) {
+    } else {
+      return Double.NaN;
+    }
+
```
Extra newline.

jnh5y commented on modules/core/src/main/java/org/locationtech/jts/geom/impl/CoordinateArraySequence.java:
```diff
@@ -178,9 +256,9 @@ public double getOrdinate(int index, int ordinateIndex)
       case CoordinateSequence.Y:  return coordinates[index].y;
-      case CoordinateSequence.Z:  return coordinates[index].z;
+      default:
+	      return coordinates[index].getOrdinate(ordinateIndex);
```
Sanity check, does one get Double.NaN or an exception if they go out of bounds?

jodygarnett replied:
found it was a little inconsistent. My understanding is exception if you are out of bounds, DnD if you are in bounds buy the value is undefined.

jnh5y commented on modules/core/src/main/java/org/locationtech/jts/geom/impl/CoordinateArraySequenceFactory.java:
```diff
@@ -78,4 +78,21 @@ public CoordinateSequence create(int size, int dimension) {
+    if (dimension < 2)
+      dimension = 2; // handle bogus dimension
+    if (dimension - measures < 2) {
+      throw new IllegalArgumentException("max spatial dimension 2 required");
```
Do you mean 'Minimum spatial dimension of 2 required'?  (I'm also a fan of dumping the inputs.  E.g. "Minimum spatial dimension of 2 required.  Input dimension was $dimension, input number of measures was $measures.")

jodygarnett replied:
fixed

jnh5y commented on modules/core/src/main/java/org/locationtech/jts/geom/impl/PackedCoordinateSequence.java:
```diff
@@ -271,14 +300,14 @@ public Double(Coordinate[] coordinates, int dimension) {
      */
     public Double(Coordinate[] coordinates) {
-      this(coordinates, 3);
+      this(coordinates, 3,0);
```
(Picky) space after the comma.

jodygarnett replied:
aside: I could not find any code formatting guidelines for the project, I setup some defaults that kind of matched in my IDE an hoped for the best.

jnh5y commented on modules/core/src/test/java/org/locationtech/jts/geom/CoordinateTest.java:
```diff
@@ -163,6 +163,82 @@ public void testDistance3D() {
+      assertEquals( xym, coord );
+      assertTrue( !xym.equalInZ(coord,0.000001) ); 
+  }
+  public void testCoordinateXYMZ() {
```
Should be  testCoordinateXYZM

jnh5y commented on modules/core/src/test/java/org/locationtech/jts/geom/impl/PackedCoordinateSequenceTest.java:
```diff
@@ -36,5 +41,103 @@ public PackedCoordinateSequenceTest(String name)
+    Coordinate coord;
+    Coordinate[] array;
+    
+    initProgression(seq);
```
Is there an easy way to share this test code with the other class?

jodygarnett replied:
Could place it in the superclass, the two tests are slightly different though. Since one store the coordinate instances (so == returns true) and one does not (so == returns false).

jnh5y replied:
Could have an override-able method around that check?

jnh5y commented on modules/core/src/test/java/org/locationtech/jts/io/WKBTest.java:
```diff
@@ -151,7 +151,7 @@ private void runWKBTest(String wkt) throws IOException, ParseException
 	private void runWKBTestPackedCoordinate(String wkt) throws IOException, ParseException {
 		GeometryFactory geomFactory = new GeometryFactory(
-				new PackedCoordinateSequenceFactory(PackedCoordinateSequenceFactory.DOUBLE, 2));
+				new PackedCoordinateSequenceFactory(PackedCoordinateSequenceFactory.DOUBLE));
```
What's the default dimension of the PackedCoordinateSequenceFactory?

jodygarnett replied:
dimension 3, measures 0, you can see in the constructors.

jnh5y replied:
Then this is a change, no?  Here we are testing getting a PCSF which creates things of dimension 2, right?

Also, a WKBReader should be able to return values which have any 2, 3, and 4 dimensions...

jodygarnett replied:
Oh here is the deal, the previous implementation of the factory was modal! So it held on to the init value `2` and would use it as a default for some of the create methods.

I removed this modal functionality since it introduced the possibility for error, especially with the static final factory implementations. Any thread could change the "default" messing up another thread.

Sot the above change is on purpose.

jnh5y replied:
+1

jnh5y commented on pom.xml:
```diff
@@ -193,7 +193,7 @@
                 <configuration>
                     <includes>
-                        <include>**/*Test.java</include>
+                        <include>**/*Test.`java`</include>
```
Does this affect anything?

jodygarnett replied:
eek! probably.

jnh5y commented:
@jodygarnett ok, I finished a read-through.  Looks pretty close.

jodygarnett commented on modules/core/src/main/java/org/locationtech/jts/geom/CoordinateXYM.java:
```diff
@@ -0,0 +1,121 @@
+  /** Default constructor */
+  public CoordinateXYM() {
+    super();
+    this.m = 0.0;
```
Default constructor is:

> Constructs a <code>Coordinate</code> at (0,0,NaN).

jodygarnett commented on modules/core/src/main/java/org/locationtech/jts/geom/CoordinateXYM.java:
```diff
@@ -0,0 +1,121 @@
+
+  public CoordinateXYM(Coordinate coord) {
+    super(coord.x,coord.y);
+    m = getM();
```
inconsistent, `0` for a spatial dimension (XY), DnD for an optional value (Z)

jodygarnett commented:
Thanks @jnh5y I will update based on your feedback, and then we really have to call it done as I do not have more time to put towards it.

Based on the above how do you feel about:

* CoordianteSequence.createCoordiante()
* Either going ahead with `(c) Vivid Solutions, and others` or reverting to `(c) Vivid Solutions`. Can you ask and let us know what to do?

jratike80 commented:
I can read "coordiante" in quite many places, like CoordianteSequence.createCoordiante(). Are they typos or something intentional?

jodygarnett commented:
typos, responding on my phone auto corrects

jnh5y commented on modules/core/src/main/java/org/locationtech/jts/geom/Coordinates.java:
```diff
@@ -0,0 +1,90 @@
+/*
+ * Copyright (c) 2016 Vivid Solutions
```
Copyright clean-up.

jnh5y commented:
@dr-jts Any more comments on this?  I'd like to merge this PR later this week if there are no understanding issues.
